### PR TITLE
serverutils: implement the testserver interfaces indirectly

### LINKED ISSUE #107058: serverutils: refactor the server test APIs
## Context

As part of our work on #76378  we are observing a pattern:

```go
s, sqlDB, kvDB := serverutils.StartServer(...)

// At this point, sqlDB and kvDB point to the "application layer" already. This is desirable!

// If a test needs something else (e.g. "AdminURL") it needs to do a bit extra work:
ts := s.TenantOrServer() // give access to the application layer.

someuse := ts.ClusterSettings()
someuse := ts.AdminURL()
// etc.
```

Meanwhile, if the test needs explicit access to the storage layer, it is cumbersome:
```go
_ = s.AdminURL() // this kinda works, but we feel it is ambiguous.
// Very cumbersome to retrieve a SQL connection:
systemSqlDB := serverutils.OpenDBConn(t, s.SQLAddr() /* don't mix with ServingSQLAddr()! */, ...)
```

## Description of problem

There are two shorcomings with this structure:
- if a test cares about the "application layer", it is too easy to **mistakenly reach out for the TestServerInterface** instead of thinking to call `TenantOrServer()` first.
- if a test cares about the storage layer / system tenant, retrieving a SQL connection is cumbersome.

And **most importantly**:

- a single run of the test only exercices EITHER the system tenant OR a secondary tenant, and never both. This halves test coverage.

### More explicit testserver APIs

At the outset, we should make the "application" vs "storage" layer separation more clear. That is:

- don't make `TestServerInterface` inherit from `TestTenantInterface` any more
- instead, make a method `StorageLayer() TestTenantInterface`  inside `TestServerInterface` so that a test can spell out explicitly that it wants access to the storage layer.
- rename `TenantOrServer()` to `ApplicationLayer()`.
- add a `SQLConn() *goSQL.DB` method to `TestTenantInterface`, it's been sorely missing!
- spell out in docs that the 2nd and 3rd return value of `StartServer` are equivalent to `.ApplicationLayer().SQLConn()` and `.ApplicationLayer().DB()`, for clarity.
- rename `TestTenantInterface` to `TestClientInterface`.

This would already clarify a great deal of tests. But it would still limit each test run to either test one or the other.

### Helper to automate working with test servers

To increase test coverage, we should **invert control** of `StartServer`, that is deprecate

```go
func StartServer(*testing.T, ...) (TestServerInterface, *sql.DB, *kv.DB)
```

and instead introduce:
```
func RunWithServer(*testing.T, testName string, ..., runFn func(*testing.T, TestServerInterface, *sqlDB, *kv.DB))
```

So a test would change from:
```go
s, sqlDB, kvDB := serverutils.StartServer(t, ...)
defer s.Stopper().Close(...)

ts := s.TenantOrServer() // give access to the application layer.
someuse := ts.ClusterSettings()
someuse := ts.AdminURL()
```

to:
```go
serverutils.RunWithServer(t, ... func(*testing.T, s ApplicationLayerInterface, sqlDB ..., kvDB ...) { 
   // No need for calling stopper explicitly.
   someuse := ts.ClusterSettings()
   someuse := ts.AdminURL()
})
```

Once we have this API in place, we can make the sub-test run for both interfaces:

```go
// in serverutils, RunWithServer:
s := startServerInternal(...)
defer s.Stopper().Close()

t.Run("system tenant", func(t *testing.T) {
   ts := s.SystemLayer()
   runFn(t, ts, ts.SQLConn(), ts.DB())
})

t.Run("secondary tenant, serverless-style", func(t *testing.T) {
   tenant := s.StartTenant(...)
   runFn(t, tenant, tenant.SQLConn(), tenant.DB())
})

t.Run("secondary tenant, dedicated-style", func(t *testing.T) {
   tenant := s.StartTenant(...)
   runFn(t, tenant, tenant.SQLConn(), tenant.DB())
})
```

Jira issue: CRDB-29864

### PR DESCRIPTION
**Pull request by knz**
First commit from #109591.
Epic: CRDB-18499
Fixes #107058.

TLDR: unit tests using TestServer/TestCluster now produce warning and
notice messages upon certain suspicious API usage.

For example:
```
$ dev test //pkg/server/storage_api -f TestStatusGetFiles
=== RUN   TestStatusGetFiles
    ...
        pkg/server/storage_api/storage_api_test_test/pkg/server/storage_api/files_test.go:46: (TestStatusGetFiles)
                NOTICE: .GetStatusClient() called via implicit interface ApplicationLayerInterface;
        HINT: consider using .ApplicationLayer().GetStatusClient() instead.
    conditional_wrap.go:193: TIP: consider replacing the test server initialization from:

            ts, db, kvDB := serverutils.StartServer(t, ...)
            // or:
            tc := serverutils.StartCluster(t, ...)
            ts := tc.Server(0)

        To:

            srv, db, kvDB := serverutils.StartServer(t, ...)
            defer srv.Stop(...)
            ts := srv.ApplicationLayer()
            // or:
            tc := serverutils.StartCluster(t, ...)
            ts := tc.Server(0).ApplicationLayer()
```

The following warnings/notices are implemented.

**ApplicationLayerInterface**

The problem here is that the implicit implementation of
`ApplicationLayerInterface` inside `TestServerInterface` is likely
misused. If/when the test server automatically starts a secondary
tenant, the *implicit* `ApplicationLayerInterface` refers to the
system tenant, whereas only `.ApplicationLayer()` refers to the
secondary tenant. It's likely that the test code should use the
latter.

For this, a warning is printed when a method from the *implicit*
`ApplicationLayerInterface` is called.

The warning is not printed if the test has indicated (via
`DefaultTestTenant: base.TestIsForStorageLayerAndNeedsASystemTenant`)
that it is ever only interested in the storage layer.

**StorageLayerInterface**

We want to promote the use of the explicit interface accessor
`.StorageLayer()` to access the methods of `StorageLayerInterface`.

For this, a notice is printed when a method from the *implicit*
`StorageLayerInterface` is called.

The notice is not printed if the test has indicated (via
`DefaultTestTenant: base.TestIsForStorageLayerAndNeedsASystemTenant`)
that it is ever only interested in the storage layer.

**TenantControlInterface**

We want to promote the use of the explicit interface accessor
`.TenantController()` to access the methods of
`TenantControlInterface`.

For this, a notice is printed when a method from the *implicit*
`TenantControlInterface` is called.

### CODE REVIEW

File: pkg/testutils/serverutils/conditional_wrap.go [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+) TestServerInterface {
+	wrapper := &wrap{
+		// The default logger function puts the info/warning messages
+		// to the standard output. This is a good default when we don't
+		// have a *testing.T.
+		// If we do, the logger function is overridden in StartServerOnlyE().
+		loggerFn: func(format string, args ...interface{}) { fmt.Printf(format, args...) },
+
+		// Used to implement the implicit TestServerController interface,
+		// and the explicit .StorageLayer() and .TenantControl() accessors.
```
Comment (stevendanna) [inline pkg/testutils/serverutils/conditional_wrap.go:55]: s/TenantControl/TenantController/

File: pkg/testutils/serverutils/fwgen/gen.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+				forwardedArgs = append(forwardedArgs, arg)
+				p("%s", arg)
+			} else {
+				for j, name := range field.Names {
+					if j > 0 {
+						p(", ")
+					}
+					id := name.Name
+					if id == "_" {
+						id = fmt.Sprintf("arg%d", argNum)
```
Comment (stevendanna) [inline pkg/testutils/serverutils/fwgen/gen.go:143]: Maybe prepend and `_` to continue to signal that this is unused.

File: pkg/testutils/serverutils/fwgen/gen.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+		p("\treturn &%s{fwAccessorFn}\n}\n\n", outy)
+	}
+	for _, m := range it.Methods.List {
+		ft, ok := m.Type.(*ast.FuncType)
+		if !ok {
+			continue
+		}
+
+		p("// %s is part of the interface %s.\n", m.Names[0].Name, itname)
+		p("func (f *%s) %s(", outy, m.Names[0].Name)
```
Comment (stevendanna) [inline pkg/testutils/serverutils/fwgen/gen.go:123]: I think we are safe for now, but I imagine this breaks from generic functions.

**Review by stevendanna — COMMENTED** [review]
Overall, I like that you've gone the extra mile to implement things like only printing the warning once per test server, an env var to enable a more verbose mode, and some facilities to test this test code.

I'm assuming we went with the code generation here to avoid having to maintain implementations of every interface function on the wrap struct manually.

If we get complaints about the noise of this, I think we could reconsider the warning on StorageLayerInterface since I don't think there is ambiguity there.

I left some very minor comments that I came across while reading through this.  I think my main concern while reading is that the code seemed a little complex looking for what it does. But, each bit of complexity does come with a reason.

**Review by knz — COMMENTED** [review]
> I'm assuming we went with the code generation here to avoid having to maintain implementations of every interface function on the wrap struct manually.

Correct.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109612#-:-NdBPO8-Bu5MBznYb_uC:b-m3o7f2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, @samiskin, and @stevendanna)

___
*[`pkg/testutils/serverutils/conditional_wrap.go` line 55 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/109612#gh-1310107589:-NdBOpPH1jvohvXF9o28:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fedab74ce848279c7cfa156a6fb281cde263f31b/pkg/testutils/serverutils/conditional_wrap.go#L55)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

s/TenantControl/TenantController/
</blockquote></details>

Done.
___
*[`pkg/testutils/serverutils/conditional_wrap.go` line 168 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/109612#gh-1310059392:-NdBP2Df-XJEtOy22hGq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fedab74ce848279c7cfa156a6fb281cde263f31b/pkg/testutils/serverutils/conditional_wrap.go#L168)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I think it would be worth a comment clarifying that these are called by the generated forwarding code.
</blockquote></details>

Done.
___
*[`pkg/testutils/serverutils/conditional_wrap.go` line 198 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/109612#gh-1310079928:-NdBP5gl6laNUats3CCr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fedab74ce848279c7cfa156a6fb281cde263f31b/pkg/testutils/serverutils/conditional_wrap.go#L198)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

s/makeBenignNotifyFn/makeSeriousNotifyFn/
</blockquote></details>

Done.
___
*[`pkg/testutils/serverutils/conditional_wrap.go` line 202 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/109612#gh-1310086441:-NdBPGKn-cv9Al3VPK-n:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fedab74ce848279c7cfa156a6fb281cde263f31b/pkg/testutils/serverutils/conditional_wrap.go#L202)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

We could mention that we require a function pointer here because of the late binding of the log function.
</blockquote></details>

Done.
___
*[`pkg/testutils/serverutils/test_server_shim.go` line 241 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/109612#gh-1310031909:-NdBPLmBCNcTFM5Mw9Pw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bab0d647de61ee84dcd9a5798b882e81ee5ac55/pkg/testutils/serverutils/test_server_shim.go#L241)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

s/ShouldStartDefaultTestTennat/ShouldStartDefaultTestTenant/
</blockquote></details>

Done.
___
*[`pkg/testutils/serverutils/fwgen/gen.go` line 1 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/109612#gh-1310061927:-NdBOHfMB3mQx78o1EfR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fedab74ce848279c7cfa156a6fb281cde263f31b/pkg/testutils/serverutils/fwgen/gen.go#L1)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

For generators like this, I think it is often useful to put a documentation comment that shows how to use it and a small example of the code it attempts to generate.
</blockquote></details>

Done.
___
*[`pkg/testutils/serverutils/fwgen/gen.go` line 123 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/109612#gh-1310122650:-NdBOJoa51ZZLE579iiG:b-8snapq) ([raw file](https://github.com/cockroachdb/cockroach/blob/fedab74ce848279c7cfa156a6fb281cde263f31b/pkg/testutils/serverutils/fwgen/gen.go#L123)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I think we are safe for now, but I imagine this breaks from generic functions.
</blockquote></details>

IMHO we're a long way from this from a go language definition perspective. Added a disclaimer in docstring.
___
*[`pkg/testutils/serverutils/fwgen/gen.go` line 143 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/109612#gh-1310115760:-NdBOhGB-iPpjbXzVaAT:b57u3wa) ([raw file](https://github.com/cockroachdb/cockroach/blob/fedab74ce848279c7cfa156a6fb281cde263f31b/pkg/testutils/serverutils/fwgen/gen.go#L143)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Maybe prepend and `_` to continue to signal that this is unused.
</blockquote></details>

discussed - all args are used.

**Review by stevendanna — APPROVED** [review]

**Comment by knz** [comment]
TFYRs!

bors r=stevendanna

**Comment by knz** [comment]
bors r=stevendanna

**Comment by knz** [comment]
bors r=stevendanna
